plugins/woocommerce/includes/class-wc-install.php

```diff
@@ -494,9 +494,10 @@ private static function setup_environment() {
 	 * @return boolean
 	 */
 	public static function is_new_install() {
-		$product_count = array_sum( (array) wp_count_posts( 'product' ) );
-
-		return is_null( get_option( 'woocommerce_version', null ) ) || ( 0 === $product_count && -1 === wc_get_page_id( 'shop' ) );
+		global $wpdb;
+		// this query will give you either 0 or 1 depending on whether a product exists or not
+		$product_exists = $wpdb->get_var("select CASE WHEN EXISTS (SELECT * from $wpdb->posts where post_type = 'product') THEN 1 ELSE 0 END as product_exists;");
```

Comment:
This is probably a lot faster than `wp_count_posts()` (if a cached value is not available), but I think we can improve it further.

```suggestion
		$product_exists = (bool) $wpdb->get_var( "SELECT DISTINCT ID FROM $wpdb->posts WHERE post_type = 'product' ORDER BY ID ASC LIMIT 1" );
```

In my testing, this has a simpler query plan and results in fewer rows being scanned (whereas, as currently written, it would seem all rows in the table are scanned in relation to the sub-query).

Reply:
Yes that's correct, both of the offending SQL statements cause a table scan and the entirety of wp_posts to be read and while it is true that this value *could* be cached, on large websites imports are often running or editors are editing/creating products or whatever, so the cache is regularly being wiped.

You can see the difference in speed on my server from this screenshot: https://www.wpintense.com/assets/image-236.png

0.0001 seconds for the way I wrote it - the version above you copied with CASE WHEN EXISTS - versus 26+ seconds for the current SQL.